File: test_step02_plot.m
close all;
clear;
clc;

addpath( 'chebfun-master' );
addpath( 'windowing_functions' );

config = F_load_json_config( 'configs/test_step02_plot.json' );

D_o = config.D_o;
alpha = config.alpha;
D_f = config.D_f;
d = config.d;

y_limit = config.y_limit;
y_number = config.y_number;
y = linspace( 0, y_limit, y_number );

t_limit = config.t_limit;
t_number = config.t_number;
t = linspace( 0, t_limit, t_number );
t = t( 2 : end );

[ Y, T ] = meshgrid( y, t );

solution_values = zeros( size( Y ) );

tic;
for i = 1 : length( t )
    FFPESolver = C_FFPESolver( d, alpha, D_o, D_f, t( i ) );
    FFPESolver.general_initialization();
    for j = 1 : length( y )
        solution_values( i, j ) = FFPESolver.get_value( y( j ) );
    end
end
toc;

set( gcf, 'Position', [ 40, 0, 1600, 1200 ] );
surf( Y, T, solution_values );
set( gca, 'ZScale', 'log' );

File: python_version/test_step02_plot.py
import argparse
import os
import sys
import time

import matplotlib.pyplot as plt
import numpy as np


CURRENT_FOLDER = os.path.dirname( os.path.abspath( __file__ ) )
REPO_FOLDER = os.path.dirname( CURRENT_FOLDER )
if CURRENT_FOLDER not in sys.path:
    sys.path.insert( 0, CURRENT_FOLDER )

from lib.config import load_config
from lib.defaults import get_local_solver_parameters
from lib.solver import FFPESolver


DEFAULT_CONFIG_PATH = os.path.join( REPO_FOLDER, 'configs', 'test_step02_plot.json' )
DEFAULT_OUTPUT_PATH = os.path.join( CURRENT_FOLDER, 'test_step02_plot.pdf' )


def ensure_directory( file_path ):
    os.makedirs( os.path.dirname( file_path ), exist_ok = True )


def build_solver_parameters( config ):
    return get_local_solver_parameters( config )


def parse_arguments():
    parser = argparse.ArgumentParser()
    parser.add_argument( '--config', type = str, default = DEFAULT_CONFIG_PATH )
    parser.add_argument( '--output', type = str, default = DEFAULT_OUTPUT_PATH )
    return parser.parse_args()


def main():
    args = parse_arguments()
    config = load_config( args.config )

    y_values = np.linspace( 0.0, float( config[ 'y_limit' ] ), int( config[ 'y_number' ] ) )
    t_values = np.linspace( 0.0, float( config[ 't_limit' ] ), int( config[ 't_number' ] ) )[ 1 : ]
    Y, T = np.meshgrid( y_values, t_values )
    solution_values = np.zeros_like( Y )

    timer = time.perf_counter()
    for i, t in enumerate( t_values ):
        solver = FFPESolver(
            d = int( config[ 'd' ] ),
            alpha = float( config[ 'alpha' ] ),
            D_o = float( config[ 'D_o' ] ),
            D_f = float( config[ 'D_f' ] ),
            t = float( t ),
            parameters = build_solver_parameters( config )
        )
        solver.general_initialization()
        for j, y in enumerate( y_values ):
            solution_values[ i, j ] = solver.get_value( float( y ) ).value
    total_seconds = time.perf_counter() - timer

    ensure_directory( args.output )
    fig = plt.figure( figsize = ( 10, 7 ) )
    axis = fig.add_subplot( 111, projection = '3d' )
    axis.plot_surface( Y, T, solution_values, cmap = 'viridis' )
    axis.set_xlabel( 'y' )
    axis.set_ylabel( 't' )
    axis.set_zlabel( 'solution' )
    fig.tight_layout()
    fig.savefig( args.output, dpi = 300 )
    plt.close( fig )

    print( 'total_seconds = {:.6f}'.format( total_seconds ) )
    print( 'output = {}'.format( args.output ) )


if __name__ == '__main__':
    main()
